core/ir/aarch64/codec.c

```diff
@@ -1484,6 +1511,55 @@ encode_opnd_imm4(uint enc, int opcode, byte *pc, opnd_t opnd, OUT uint *enc_out)
     return encode_opnd_int(8, 4, false, 0, 0, opnd, enc_out);
 }
 
+/* cmode4_s_sz_msl: Operand for 32 bit elements' shift amount (shifting ones) */
+
+static inline bool
+decode_opnd_cmode4_s_sz_msl(uint enc, int opcode, byte *pc, OUT opnd_t *opnd)
+{
+    /* cmode size amounts
+     * 1110  8    0
+     * 10x0  16   0,8
+     * 0xx0  32   0,8,16,24
+     * 110x  32   8,16 (MSL, shifting ones)
+     */
+    int cmode4 = extract_uint(enc, 12, 1);
+    int esize = 32;
+    int shift = ((cmode4 == 0) ? 8 : 16) | (1 << 28);
+    uint64 sz_shft = ((uint64)esize << 32) | shift;
+    *opnd = opnd_create_immed_int64(sz_shft, OPSZ_8);
+    return true;
+}
+
+static inline bool
+encode_opnd_cmode4_s_sz_msl(uint enc, int opcode, byte *pc, opnd_t opnd,
+                            OUT uint *enc_out)
+{
+    if (!opnd_is_immed_int64(opnd))
+        return false;
+
+    int64 sz_shft = opnd_get_immed_int64(opnd);
+    int shift = (int)(sz_shft & 0xFFFFFFFF);
+    if ((shift & 0x10000000) == 0) // MSL bit
+        return false;
+    shift = shift & 0xFF;
+    int esize = (int)(sz_shft >> 32);
```

Comment:
Wait: you're storing the element size in the upper bits?  Is that an intra-DR decision and unrelated to the ISA?  Please document this.  So we're taking two different integers and merging them to store as a single operand?  Would it be simpler to use two separate operands, and that would avoid this int64 opnd type?

Reply:
Yes, agreed. This is a one-off for this particular type of encoding and temporary. Further work is required on the codec generator to be able to specify different numbers of operands and will form part of upcoming development work on codec generator improvements and preparing for future work.

Reply:
Added comments which explain size/shift amount operand and reference to #4393.